src/Strategy/Go/GoListPackages.hs

```diff
@@ -0,0 +1,240 @@
+        -- Not doing this would make us do static analysis on projects that previously worked with the `go mod graph` based tactic.
+        -- See https://github.com/kubernetes/kubernetes/tree/master/hack/tools for a project needing this behavior.
+        -- See https://pkg.go.dev/cmd/go/internal/list for documentation of this flag
+        , "-json" -- Output in json format
+        , "-deps" -- Recursively print deps
+        , "all" -- all packages
+        ]
+    , cmdAllowErr = Never
+    }
+
+analyze ::
+  ( Has Exec sig m
+  , Has Diagnostics sig m
+  ) =>
+  Path Abs Dir ->
+  m (Graphing.Graphing Dependency, GraphBreadth)
+analyze goModDir = do
+  stdout <- context ("Getting dependencies using '" <> renderCommand goListCmd <> "'") $ execThrow goModDir goListCmd
+  case decodeMany stdout of
+    Left (path, err) -> fatal $ CommandParseError goListCmd (toText (formatError path err))
+    Right pkgs -> do
+      context "Analyzing dependencies" $ buildGraph pkgs
+
+-- | Given a list of GoPackages, generate a graph of *module* dependencies based on their package dependencies.
+-- This works by first making a mapping of package import paths to modules using the provided packages' 'moduleInfo' field.
+-- Next each package's deps have edges made between a it and its dependencies, except that the function uses a package's parent module to generate a the vertex for the package in the graph.
+-- The end result is a graph of module dependencies for only modules that have at least one of its packages used by the built project.
+--
+-- Other features of this function:
+-- * Removes standard lib deps where they are called out by a flag.
+--   Stdlib packages with pseudo versions are not filtered yet. This is the same behavior as the current dynamic tactic in Strategy.Go.GoList.
+-- * Removes path dependencies and child deps that aren't used elsewhere in the graph.
+--   The go tools give us this data but we haven't decided yet how to present it.
+-- * Replaces modules according to the module's 'replacement' field.
+buildGraph :: (Has Diagnostics sig m) => [GoPackage] -> m (Graphing.Graphing Dependency, GraphBreadth)
+buildGraph rawPackages =
+  fmap ((,Complete) . pruneUnreachable) -- Remove deps that are only children of removed path deps.
+    . evalGrapher
+    . traverse_ graphEdges
+    $ pkgsNoStdLibImports
+  where
+    (stdLibImportPaths, pkgsNoStdLibImports) = foldl' go (HashSet.empty, HashMap.empty) rawPackages
+
+    go :: (HashSet.HashSet ImportPath, HashMap.HashMap ImportPath GoPackage) -> GoPackage -> (HashSet.HashSet ImportPath, HashMap.HashMap ImportPath GoPackage)
+    go (stdLibPaths, otherPackages) g@GoPackage{standard, importPath}
+      | standard = (HashSet.insert importPath stdLibPaths, otherPackages)
+      | otherwise = (stdLibPaths, HashMap.insert importPath (removeIgnoredPackages g) otherPackages)
```

Comment:
I may be totally missing something here, I'm guessing this is recursive and I'm missing where that recursion is happening, but how is it possible to use `removeIgnoredPackages` here inside `go` if `removeIgnoredPackages` uses `ignoredPackages` which uses `stdLibImportPaths` which depends on the output of `foldl' go`? Is this not a circular dependency?

Reply:
Talked offline and this makes sense now!